### PR TITLE
feat(install): make install script posix compliant

### LINKED ISSUE #2226: feat(install): install script should not require bash
#### Is your feature request related to a problem? Please describe.
The install.sh script currently requires bash. Not all linux distributions include bash by default and there is no reason the script could not be posix compliant (it looks like it mostly is already).

#### Describe the solution you'd like
Make the install script posix complaint and use /usr/bin/env sh for the shebang

#### Describe alternatives you've considered
Installing bash as a prerequisite works, but should not be necessary

NOTE:

I am happy to take on this work and will submit a PR if the community / maintainers agree.

### PR DESCRIPTION
**Pull request by dmccaffery**
#### Description

* update shebang to support posix-compliant shells (like dash)
* replace all usages of echo with printf (to normalise)
* remove locals, which are not posix complaint
* rename `command` variable to `fetch_cmd` as it overloads a posix builtin
* rename `complete` function to `completed` as it overloads the bash builtin for completion

#### Motivation and Context

Bash should not be necessary to run the install script on systems that would otherwise not require it. Ubuntu uses dash by default and I use zsh, so no reason to install bash before installing starship. :)

Closes #2226 

#### How Has This Been Tested?

Ran install scripts on macOS natively, as well as the following docker containers for linux distributions (not the same as native, but as close as I can get with ease):

* debian
* fedora
* centos
* linuxmint
* ubuntu

- [X] I have tested using **MacOS**
- [X] I have tested using **Linux**
- [x] I have tested using **Windows**

#### Checklist:

- [ ] (not relevant) I have updated the documentation accordingly.
- [ ] (not relevant) I have updated the tests accordingly.

### CODE REVIEW

**Comment by davidkna** [comment]
I don't really see the value of removing `local` just for the sake of POSIX compliance, because I can't think of any shell that doesn't support it. Even `busybox` has support for it. ([More elaborate comment on stackoverflow](https://stackoverflow.com/a/18600920)).

File: install/install.sh [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!/bin/bash
+#! /usr/bin/env sh
```
Comment (davidkna) [inline install/install.sh:1]: Is the space here intentional? I think some systems could have issues with that.

File: install/install.sh [hunk 2]
```diff
@@ -20,18 +20,18 @@
 #   -B, --base-url
 #     Override the base URL used for downloading releases
 
-set -euo pipefail
```
Comment (davidkna) [inline install/install.sh:23]: Isn't pipefail also nearly universally supported? It might be a bit worse than `local`, I guess.

Comment (davidkna) [reply install/install.sh:23]: I think `set -o pipefail 2>/dev/null || true` in a separate line should be POSIX compliant.

**Comment by dmccaffery** [comment]
Agreed. I removed them because I don't think dash supports it (default shell in Ubuntu), but will verify.

That being said, they are only global in process and the install script is invoked; not evaluated.

File: install/install.sh (hunk 1 above)
Comment (dmccaffery) [reply install/install.sh:1]: It was intentional and I always do it, but I never knew it could be an issue. Apparently it's a great debate: https://web.archive.org/web/20170215070707/http://www.in-ulm.de/~mascheck/various/shebang/#details

I have no issue with removing the space; and will do!

File: install/install.sh (hunk 2 above)
Comment (dmccaffery) [reply install/install.sh:23]: I can add this; it should be okay and will verify. It does fail in dash.

File: install/install.sh (hunk 1 above)
Comment (dmccaffery) [reply install/install.sh]: done

File: install/install.sh (hunk 2 above)
Comment (dmccaffery) [reply install/install.sh:23]: done

File: install/install.sh (hunk 1 above)
Comment (dmccaffery) [reply install/install.sh]: `set -o pipefail` immediately exists regardless of any attempt to ignore the exit code with `|| true` in dash... simple test:

```sh
temp=$(mktemp -d)
cat << EOT > "$temp"/with-pipefail.sh
#! /usr/bin/env sh

set +e
printf "%s\n" "attempting to set pipefail"
set -o pipefail 2>/dev/null || true
printf "%s\n" "we got here"
EOT

chmod +x "$temp"/with-pipefail.sh

docker run --volume="$temp":/test debian /test/with-pipefail.sh
rm -rf "$temp"
```

**Comment by dmccaffery** [comment]
I just realised that modern versions of curl will fail with binary files when not used with a pipe or `--output`. I tried to use `--output -` to no avail; not sure how to resolve without dramatically modifying the fetch logic to not write to the stdout buffer.

**Comment by vladimyr** [comment]
I guess that is time to close #1315

**Comment by davidkna** [comment]
@vladimyr Maybe. I recently duplicated a lot of your changes from there so merge conflicts are likely significant. But if you still want to get any of the other changes from that PR merged please rebase it.

@dmccaffery Would you please port over `download` and `unpack` from #1315 to make sure error handling without `pipefail` still works?

**Comment by dmccaffery** [comment]
@davidkna thanks and will do -- apologies for the delay; I've been busy.

@vladimyr -- I'm so sorry I didn't see your PR before opening this one. I did a search for an issue but didn't think to scan current PRs!

**Comment by dmccaffery** [comment]
@davidkna I made the requested changes and also added a new GitHub Actions workflow for testing the install across a wide array of environments, including Windows WSL. This is isolated on the last commit in case you do not want to incorporate it.

Let me know if there is anything else you need!

File: .github/workflows/install.yml [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+jobs:
+  check:
+    name: check
+    runs-on: ubuntu-20.04
+    container: koalaman/shellcheck-alpine:stable
+    steps:
+      - name: checkout
+        uses: actions/checkout@v2
+      - name: shell
+        run: shellcheck -e SC1090,SC1091,SC1071 install/install.sh
```
Comment (davidkna) [inline .github/workflows/install.yml:19]: There's a comment in `install.sh` to disable `SC2039`. Why are these warning disabled  only here?

File: install/build.sh [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+	--build-arg IMAGE="$IMAGE" \
+	--label=org.opencontainers.image.name=starship \
+	--label=org.opencontainers.image.revision="$GITHUB_SHA" \
+	--label=org.opencontainers.image.version="$GITHUB_REF" \
+	--label=org.opencontainers.image.source="https://github.com/$GITHUB_REPOSITORY/starship" \
+	--tag "ghcr.io/$GITHUB_REPOSITORY/starship:$TAG" \
+	"${PWD}/install"
+
+    # this can be added if images are ever published
+	# --cache-from=type=registry,ref="ghcr.io/$GITHUB_REPOSITORY/starship:$IMAGE-latest" \
```
Comment (davidkna) [inline install/build.sh:75]: Is there a use-case to distribute Starship in a container?

**Review by davidkna — CHANGES_REQUESTED** [review]

**Comment by davidkna** [comment]
I think it might be better to move the testing changes to a separate PR so this can be merged quickly.

File: install/build.sh (hunk 4 above)
Comment (dmccaffery) [reply install/build.sh:75]: Not that I can think of except to be able to quickly pull one to reproduce an issue someone reports with a specific flavour combination. I can't think of any reason beyond that.

File: .github/workflows/install.yml (hunk 3 above)
Comment (dmccaffery) [reply .github/workflows/install.yml:19]: 1090 and 1091 have to do with spellcheck being able to follow sourced or exec'd scripts, which it can't do when things are executed from a url, download, or uncompressed asset. These could be moved into the script, but it's fairly common to disable these for scripts that bootstrap other scripts over the wire.

1071 isn't actually necessary here, but it "ignores" scripts that shellcheck can't lint based on the shebangs (like zsh, for example). We don't have any of those in this use case.

It's a style choice and I'm happy to go either way.

File: install/install.sh [hunk 5]
```diff
@@ -104,41 +110,69 @@ fetch() {
...
+    return 1
   fi
+
+  $cmd && return 0 || rc=$?
+
+  error "Command failed (exit code $rc): ${BLUE}${cmd}${NO_COLOR}"
+  printf "\n" >&2
+  info "This is likely due to Starship not yet supporting your configuration."
+  info "If you would like to see a build for your configuration,"
+  info "please create an issue requesting a build for ${MAGENTA}${arch}-${platform}${NO_COLOR}:"
```
[the remark this answers is not in the document]
Comment (dmccaffery) [reply install/install.sh:144]: Yes. Will do.

File: .github/workflows/install.yml [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+          - ubuntu-16.04
+    steps:
+      - name: checkout
+        uses: actions/checkout@v2
+      - name: install
+        run: ./install/install.sh --force
+      - name: test
+        run: |
+          starship --version
+          command -v starship
```
[the remark this answers is not in the document]
Comment (dmccaffery) [reply .github/workflows/install.yml:42]: Can do. Prefer this to all be containerised, or on Ubuntu?

**Comment by dmccaffery** [comment]
Agreed. I'll pull the testing out and open a separate PR.

File: .github/workflows/install.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (davidkna) [reply .github/workflows/install.yml:42]: Just ubuntu is fine but macOS would be nice too, especially given the old bash version it ships with.

File: .github/workflows/install.yml (hunk 3 above)
Comment (davidkna) [reply .github/workflows/install.yml:19]: In that case a comment in your future PR would be nice.

File: install/install.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (dmccaffery) [reply install/install.sh]: done

**Comment by dmccaffery** [comment]
@davidkna : removed the install tests, which have been reopened in a WIP/draft PR for now: #2406 -- also made the requested change to the download function, so should be ready for a final review. Thanks!

File: install/install.sh [hunk 7]
```diff
@@ -189,7 +199,7 @@ detect_platform() {
     freebsd) platform="unknown-freebsd" ;;
   esac
 
-  echo "${platform}"
+  printf '%s\n' "${platform}"
```
[the remark this answers is not in the document]
Comment (dmccaffery) [reply install/install.sh]: echo emits a newline, which is why I kept it in printf -- has no side effects to change, though so accepted recommendation.

**Review by davidkna — APPROVED** [review]
LGTM

**Comment by chipbuster** [comment]
@dmccaffery Thank you for your contributions, and thanks @davidkna for reviewing!

**Comment by vladimyr** [comment]
> @vladimyr -- I'm so sorry I didn't see your PR before opening this one. I did a search for an issue but didn't think to scan current PRs!

@dmccaffery No worries I didn't post my comment to criticize but to spare you some time dealing with conversion. I hope it was helpful and you found useful stuff over there :crossed_fingers: 

> @vladimyr Maybe. I recently duplicated a lot of your changes from there so merge conflicts are likely significant.

@davidkna Let's put it this way. I'm in no way emotionally attached to my code but what is the point of duplicating? You could simply continue on top of it or at least copy it completely. Doing neither leads to the omission of important stuff like:

> - support `PREFIX` environment variable
> 
> Added support for `PREFIX` environment variable which enables the following _standardish_ invocation style:
> ```sh
> curl -fsSL https://starship.rs/install.sh | PREFIX=$HOME/.local sh
> # starship ends up inside $HOME/.local/bin
> ```

discussed in detail here: https://github.com/starship/starship/pull/1315#discussion_r439212699

But yeah, you gave me an _interesting_ option:

> But if you still want to get any of the other changes from that PR merged please rebase it.

:zipper_mouth_face: ...

Let's get back to the technical track...

@davidkna said:
> I don't really see the value of removing `local` just for the sake of POSIX compliance, because I can't think of any shell that doesn't support it. Even `busybox` has support for it. ([More elaborate comment on stackoverflow](https://stackoverflow.com/a/18600920)).

1. SO user literally mentioned `ksh` and non-GNU/non-BSD systems as potential deal-breakers:
	  > So unless you're aiming to support **non-GNU non-BSD systems** like Solaris, or those using **standard ksh**, etc., **you could get away with using local**. (Might want to put some comment right at the start of the script, below the shebang line, noting that it is not strictly a POSIX sh script. Just to be not evil.)
	  
	  Even though I do agree it is very unlikely to get a bug report from the latter I wouldn't be so surprised to see a `ksh` user around.

2. The fact that you _could get away with it_ doesn't mean you should! To prove my point let me show you Docker's install script: https://get.docker.com Try to find some `local`s inside :wink: 

3. There are reasons why `local` isn't standardized by POSIX and while following that [debian.org discussion link](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=381091) gives you some insight here is the link to the actual Open Group bug tracker https://www.austingroupbugs.net/bug_view_page.php?bug_id=767 
	Once you scratch under the surface you figure out that the answer to _Are there any shells out there that don't support `local`?_ is not that important because that is not the right question to ask in the first place. As this SE user pointed out:
	> It's not as simple as supporting `local` or not.
	
	https://unix.stackexchange.com/questions/493729/list-of-shells-that-support-local-keyword-for-defining-local-variables
	
4. Regardless of what you end up doing disabling SC2039 is **wrong** thing to do! First of all, it is an umbrella warning so you just disabled a whole bunch of other helpful warnings plus that is the exact reason why it is deprecated:
	> _Note: This warning has been retired in favor of individual SC3xxx warnings for each individual issue._
	
	https://github.com/koalaman/shellcheck/wiki/SC2039 Unfortunately, that change is relatively new (https://github.com/koalaman/shellcheck/wiki/SC2039/_compare/e241394...6c9fa14 & https://github.com/koalaman/shellcheck/wiki/SC2039/6c9fa14) so those `SC3xxx` rules haven't been implemented yet. So you either live with shellcheck yelling at you or you trade safety for commodity.
	
Finally, regardless of all that thanks both of you for your contribution and for moving things forward! :+1:

**Comment by vladimyr** [comment]
FWIW Once the new ShellCheck build gets released we could disable `SC3043` only :tada: 
https://github.com/koalaman/shellcheck/blob/81acea0/src/ShellCheck/Checks/ShellSupport.hs#L339-L345

**Comment by vladimyr** [comment]
> I wouldn't be so surprised to see a `ksh` user around.

To further strengthen my argument about `ksh`'s `local` support let's take a look at the following. I have Termux installed on my phone and I'm using `zsh` with starship OFC:
```
~
❯ echo $ZSH_VERSION
5.8

~
❯ type local
local is a reserved word
```

Switching quickly to `bash` gives the following output:
```
~ $ echo $BASH_VERSION
5.1.4(1)-release
~ $ type local
local is a shell builtin
```

But Android (11) ships with [`mksh` (MirBSD Korn Shell)](http://www.mirbsd.org/mksh.htm) (R57): https://android.googlesource.com/platform/external/mksh/+/refs/heads/android11-release/src/sh.h#187
So let's switch again to `/system/bin/sh` to check its `local` support:
```
P2a42:/ $ echo $KSH_VERSION
@(#)MIRBSD KSH R52 2016/01/20
P2a42:/ $ type local
local is an alias for '\typeset'
```

Apart from the fact it reveals I'm using a bit older Android OS/device :upside_down_face:, this is still a fairly modern `ksh` clone and yet `local` is defined as an alias! :warning:

**Comment by davidkna** [comment]
> What is the point of duplicating?
 
I wasn't aware of your PR.

> I wouldn't be so surprised to see a `ksh` user around.

For the use here (locals in functions) it looks like the implementation differences don't matter at all. The only issue is standard ksh which doesn't support `local` in posix functions and I agree it would be worthwhile to support that shell.

**Comment by dmccaffery** [comment]
@davidkna / @vladimyr :

Should I remove the locals again? I don't think it will have any side effects. No one should be sourcing the install script. When executed, the scope is the forked process of the sub shell, so there is no need for locals anyway (as far as I know).

**Comment by vladimyr** [comment]
> I wasn't aware of your PR.

Ah, too bad you missed it. However, one way or another installer is finally (almost) POSIX compliant :tada: 

> The only issue is standard ksh which doesn't support `local` in posix functions and I agree it would be worthwhile to support that shell.


> Should I remove the locals again? I don't think it will have any side effects. No one should be sourcing the install script. When executed, the scope is the forked process of the sub shell, so there is no need for locals anyway (as far as I know).

It seems we have a consensus here. `ksh` is a worthwhile target to support and, as @dmccaffery nicely explained, ensuring local scope isn't that crucial under terms of preferred/advertised use case. So, @dmccaffery I'm sorry for expanding your todo list :see_no_evil: but feel free to remove locals :wink:

**Comment by dmccaffery** [comment]
... sure, but imma smack you. 😈

**Comment by vladimyr** [comment]
As of writing this comment Starship's install instructions (https://github.com/starship/starship/blob/4925764/docs/README.md#install-latest-version) suggest the following:
```sh
curl -fsSL https://starship.rs/install.sh | bash
```

First, this needs to get changed to (at least):
```sh
curl -fsSL https://starship.rs/install.sh | sh
```
in order to advertise reaching POSIX compliance to the end-users.

IMHO there are 4 likely types of workflows users may follow to install Starship using the installer script:

1. **simple (c/p) workflow** (most likely)
	In this workflow, the value of shebang is ignored as script content is piped directly to the (hopefully POSIX compatible) `sh`.

2. **advanced workflow**
	In this workflow, the user makes sure that the script is fully downloaded before starting its execution to prevent possible [timing attacks](https://unix.stackexchange.com/a/339276).
	```sh
	sh -c "$(curl -fsSL https://starship.rs/install.sh)"
	```
	Shebang is ignored in this case because the desired shell (`sh`) is explicitly invoked.
	**It is worth noting that this form should be preferred over the currently advertised one because it is equally convenient (c/p-able) but more secure.** For example, take a look at Homebrew installation docs: https://brew.sh 

3. **cautious workflow**
	In this workflow, the user first downloads the installer script:
	```sh
	curl -fsSL https://starship.rs/install.sh -o install.sh
	```
	They then verify its contents, and eventually run it:
	
	- by passing it as shell parameter
		```sh
		sh install.sh
		```
		Shebang is ignored in this case because the desired shell (`sh`) is explicitly invoked.
		**Again, it is worth noting that some projects prefer writing out these steps in detail instead of advertising simple c/p-able shell snippets thus encouraging good safety practices!** For example, here is how Docker folks do it: https://docs.docker.com/engine/install/debian/#install-using-the-convenience-script
	
	- as if it's an executable
		```sh
		chmod +x install.sh
		./install.sh
		```
		In this case, the program loader parses the shebang and uses it to spawn a shell that'll be used for executing script contents.
	
So, only in 1 out 4 possible workflows shebang actually comes into the play. In all other workflows, it serves as a mere indicator of POSIX compliance so the user knows which shell to pick/invoke. In the former case, loader decides which shell will get invoked by parsing the shebang which brings me to my point about @dmccaffery's shebang choice https://github.com/starship/starship/blob/4925764/install/install.sh#L1:
```sh
!#/usr/bin/env sh
```
IMO it is somewhat superfluous. There is hardly any advantage in using `env` and looking up `sh` from the `PATH` instead of going with plain old `/bin/sh`. Yes, I'm aware that Solaris' `/bin/sh` is actually a Bourne Shell which is not POSIX compliant and what [POSIX standard](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/sh.html#:~:text=Applications%20should%20note%20that%20the%20standard,(for%20example%2C%20with%20sed).%20For%20example%3A) says about it but bear in mind that:

1. it is part of **informative** section
2. it is safe to assume Solaris folks are typically advanced users and well aware of that quirk and they'll probably invoke the correct (POSIX compliant) shell directly anyway

Honestly, I never saw anyone suggesting the use of `env` for POSIX `sh` lookup but apparently [Wikipedia authors think it is a good idea](https://en.wikipedia.org/wiki/Shebang_(Unix)#Program_location:~:text=A%20similar%20problem%20exists%20for%20the,Character). I definitely don't share their opinion because while it makes it more foolproof it also makes it less secure too. There is no security without knowledge. If the user treats shebang as some sort of magic and rejoices at the perceived flexibility from `env` usage they are certainly also not aware that a simple tweak to their shell profile (specifically `PATH` value) together with the introduction of fake `sh` will end up with them having a very very bad day.

So, long story short, I'm proposing the following:

1. **Let's simplify things by using simple `/bin/sh` shebang.** (As we did before with `/bin/bash` too.)
2. **Let's strike a balance between safety and convenience** by advertising the following form instead:
	```sh
	sh -c "$(curl -fsSL https://starship.rs/install.sh)"
	```
